osbenchmark/synthetic_data_generator/synthetic_data_generator.py

```diff
@@ -6,98 +6,148 @@
+        self.logger.info("Number of workers to use: [%s]", workers)
+
+        console.println(f"[NOTE] ✨ Dashboard link to monitor processes and task streams: [{dask_client.dashboard_link}]")
+        console.println("[NOTE] ✨ For users who are running generation on a virtual machine, consider tunneling to localhost to view dashboard.")
+        console.println("")
+
+        self.logger.info("Average document size in bytes: [%s]", avg_document_size)
+        self.logger.info("Chunk size: [%s] docs", docs_per_chunk)
+        self.logger.info("Total GB to generate: [%s]", self.sdg_metadata.total_size_gb)
+        self.logger.info("Max file size in GB: [%s]", generation_settings.get('max_file_size_gb'))
+
+        console.println(f"Total GB to generate: [{self.sdg_metadata.total_size_gb}]\n"
+                        f"Average document size in bytes: [{avg_document_size}]\n"
+                        f"Max file size in GB: [{generation_settings.get('max_file_size_gb')}]\n")
+
+        start_time = time.time()
+        with tqdm(total=total_size_bytes,
+                    unit='B',
+                    unit_scale=True,
+                    bar_format="{l_bar}{bar}| {n_fmt}/{total_fmt} [{elapsed}<{remaining}, {rate_fmt}]") as progress_bar:
+
+            setup_custom_tqdm_formatting(progress_bar)
+            while current_size < total_size_bytes:
+                file_path = os.path.join(self.sdg_metadata.output_path, f"{self.sdg_metadata.index_name}_{file_counter}.json")
+                file_size = 0
+                docs_written = 0
+
+                while file_size < max_file_size_bytes:
+                    generation_start_time = time.time()
+                    # Generate data across all workers
+                    seeds = self.generate_seeds_for_workers(regenerate=True)
+                    self.logger.info("Using seeds: %s", seeds)
+
+                    futures = self.strategy.generate_data_chunks_across_workers(dask_client, docs_per_chunk, seeds)
+
+                    writing_start_time = time.time()
+                    for _, data in as_completed(futures, with_results=True):
+                        docs_written_from_chunk, written_bytes = write_chunk(data, file_path)
+                        docs_written += docs_written_from_chunk
+                        current_size += written_bytes
+                        progress_bar.update(written_bytes)
+                    writing_end_time = time.time()
+
+                    generating_took_time = writing_start_time - generation_start_time
+                    writing_took_time = writing_end_time - writing_start_time
+                    self.logger.info("Generating took [%s] seconds", generating_took_time)
+                    self.logger.info("Writing took [%s] seconds", writing_took_time)
+
+                    file_size = os.path.getsize(file_path)
+                    # If it exceeds the max file size, then append this to keep track of record
```

Comment:
Is the file truncated to keep under the max, or is there some leeway allowed?

Reply:
There's some leeway allowed for now. Was going to see what users prefer after they use it in 2.0